### ISSUE TITLE
Reproducible builds for Naive

### ISSUE THREAD

**Issue by klzgrad** [SagerNet/SagerNet#109]
Chromium supports reproducible builds, which mean the built output will be binary identical given the same source code. This offers some security benefit that allows third parties to verify build integrity. In this case, the Android builds of NaiveProxy can be verified using sha256 sums between multiple independent builders (though it's possible to have different sums if the Android NDK version is different).

I have updated my workflow so it will produce an artifact containing the sha256 sums of all platforms, https://github.com/klzgrad/naiveproxy/actions/runs/868812522. When the user downloads naive-plugin apk from here, he should be able to verify the libnaive.so binary using the checksums from upstream artifacts. Or the upstream builds can be cross checked using build checksums from here.

Workflow artifacts instead of release attachments are used to store the checksums because release attachments can be edited by project collaborators, so in theory I can put any rogue stuff there after the build. But workflow artifacts seem to be immutable once the workflow runs from a certain commit. If the GitHub infrastructure can be trusted, then the integrity of the checksum artifacts can be trusted, then the integrity of the binaries can be verified across projects to ensure there is no supply chain attack. The trust model here is this: the GitHub infra is trusted, the source code of a given version is trusted assuming adequate review, individual project members are not trusted as they can manually upload arbitrary release attachments at a arbitrary later date (this condition precludes use of GPG signatures). 

The only limitation is artifacts are saved only for 90 days. But I can't find another trusted platform for storing immutable data.

In summary, I'm asking for SagerNet to also produce checksums for Naive plugin builds so they can be checked against upstream checksums. I'm also open to better idea of verifying integrity of release binaries.

**Comment by nekohasekai** [comment]
https://gitlab.com/fdroid/fdroiddata/-/merge_requests

I'm submitting SN and all plugins to F-Droid, they promise the APKs they sign is the same as the source code.

The downside is that it takes a week or more to release a version update each time.

**Comment by klzgrad** [comment]
> https://gitlab.com/fdroid/fdroiddata/-/merge_requests/9044/diffs

Side notes: qemu-user is not needed if you don't run tests. ccache is not needed if no cache is saved across builds to make use of ccache results.

F-Droid has reproducible builds, okay. But it is not *verified* until a verification server runs to verify it, which is not trivial to set up. What I'm asking is that the libnaive.so reproduced by F-Droid should have the same checksum as the ones built on GitHub, so this single file is reproducible from the perspective of the Chromium build system, not F-Droid which is much larger. 

I can already reproduce Naiveproxy's GitHub builds and compare checksums locally with relative ease. So hopefully it is not too much work doing the same in F-Droid.

1. To have reproducible builds you have to revert the previous optimization of reusing Android NDK and let get-clang.sh download the NDK version corresponding to the Chrome version (r22 now, not r21e, it may change as Chrome updates.).
2. The user can run local builds to verify Naiveproxy's GitHub builds.
3. The user can download and extract SagerNet's Github apk builds, and verify the checksum of libnaive.so against Naiveproxy upstream checksums.
4. The user can download and extract F-Droid's SagerNet apk, and verify the checksum of libnaive.so against Naiveproxy upstream checksums.

"The user" includes me, of course.

**Comment by nekohasekai** [comment]
> But it is not verified

You mean, F-Droid is untrustworthy?

> The only limitation is artifacts are saved only for 90 days.

Maybe the hash value can be used as the artifact name

**Comment by klzgrad** [comment]
F-Droid should have the same level of trust as Github but that's besides the point. The point of reproducible build is to actually have working workflows that consistently reproduce the build, thus to have verified its reproducibility. Preferably the workflows run on different independent platforms as they are less likely to be simultaneously compromised.

And there are several levels of reproduction. Reproducing the full F droid apk requires more resources. The more risky part is where you introduce third party plugins and their supply chain risks. Reproducible builds are better for this type of risks.

**Comment by klzgrad** [comment]
https://github.com/klzgrad/naiveproxy/actions/runs/875211546 The hashes are generated with NDK r22 (pulled in get-clang.sh) and the naive binary only.

**Comment by klzgrad** [comment]
NaiveProxy upstream v91.0.4472.77-1 has this hash for Android ARM64:
* android-arm64-sha256 36594204283a65faf70d4ae7f542a32082486474540def45de8643d1108c93b1.
This is the hash of the `naive` binary.

It matches none of the hashes produced here. 
* https://github.com/SagerNet/SagerNet/actions/runs/887425026 "NaïveProxy Plugin Release Build": SHA256-ARM64 c66f148975407b7fc9831a5f859c8c813921fc5053b18736f17528f34922a862. This seems to be the hash of naive-plugin-91.0.4472.77-1-arm64-v8a.apk. Ok, but it can't be used to match against upstream.
* `naive-plugin-91.0.4472.77-1-arm64-v8a.apk/lib/arm64-v8a/libnaive.so` has this sha256sum: df58c175c629758b32c98123fd73751ee887904615edcdadba4250668872f227. This one should match the upstream hash.
* https://github.com/SagerNet/SagerNet/actions/runs/887345681 "Update naive to 91.0.4472.77-1": SHA256-ARM64 42ade24b8efba7a9ea51274d8da541b78d86e0c9d329598dac13433ca268b816. I have no idea what this hash is about.

**Comment by nekohasekai** [comment]
> but it can't be used to match against upstream.

It seems that the cause lies in the different NDKs. 

The question here is, if you trust GitHub, why do you need to check every lib if the APK is trusted? 

> I have no idea what this hash is about.

Useless

**Comment by nekohasekai** [comment]
Perhaps downloading the lib directly from your release is the solution we need.

**Comment by klzgrad** [comment]
The question is I don't trust myself (i.e. my opsec), so I ask for more reproduced builds, instead of reusing builds.

Can you revert this https://github.com/SagerNet/SagerNet/commit/d39569266d80e3d75c44cfec381e5f06f8bd88b5 and see if it reproduces. The additional cost should be abot 30-90 seconds.

**Comment by nekohasekai** [comment]
https://github.com/SagerNet/SagerNet/actions/runs/888430261

**Comment by klzgrad** [comment]
OK, thank you.

**Comment by nekohasekai** [comment]
https://github.com/SagerNet/SagerNet/releases/tag/naive-plugin-91.0.4472.77-1

This new release may be OK.

**Comment by klzgrad** [comment]
I don't like seeing this https://github.com/SagerNet/SagerNet/commit/21ce4891f4a69f20d46ef0b835baf1c5d95b6c8a. It went back to the non reproducing way. I checked the fdroid build log on gitlab but it seems cloning the Android ndk wasn't the most costly step. Is it this error?
```
git clone --depth=1 --filter=blob:none --no-checkout https://chromium.googlesource.com/android_ndk.git third_party/android_ndk
error: unknown option `filter=blob:none'
```
I can help fix this if this is the issue you were reverting for. 
What is the build environment of fdroid?

There are also timeout errors. Probly don't build x86 for fdroid?

**Comment by nekohasekai** [comment]
>  What is the build environment of fdroid?

Seems to be Debian stretch

> There are also timeout errors. Probly don't build x86 for fdroid?

No, this is just a limitation of GitLab CI

**Comment by nekohasekai** [comment]
Maybe fixed in [2ec99502](https://gitlab.com/nekohasekai/fdroiddata/-/commit/2ec9950235b5dc287e7466ba7452f351dd0c3e3c)

**Comment by klzgrad** [comment]
I see your new solution now. It does seem to be the only way.

My usual trick is to use debs from newer releases, but git 2.29 from buster-backports would require libc6 2.28 which is not compatible with stretch's libc6. Using older git commands would also make it very slow as even the direct download https://chromium.googlesource.com/android_ndk/+archive/401019bf85744311b26c88ced255cd53401af8b7.tar.gz takes 5 minutes. This partial sparse cloning feature is really needed to keep it under control.
